src/context/node.js

```diff
@@ -0,0 +1,414 @@
+    while (n) {
+      // Check if a node is a candidate to be returned.
+      if (n != node || !excludeSelf) {
+        if (n[NODE_PROP]) {
+          // Already a discovered node.
+          return /** @type {!ContextNode} */ (n[NODE_PROP]);
+        }
+        const {nodeType} = n;
+        if (nodeType == DOCUMENT_NODE || nodeType == FRAGMENT_NODE) {
+          // A context node is always created for a root. Due to this, a
+          // non-root element is always at least attached to a root. This
+          // allows for quick discovery and reattachment when new information
+          // becomes available.
+          return ContextNode.get(n);
+        }
+        if (nodeType == ELEMENT_NODE && startsWith(n.tagName, AMP_PREFIX)) {
+          // An AMP node will always have a context node backing it at some
+          // point.
+          return ContextNode.get(n);
+        }
+      }
+      // Navigate up the DOM tree. Notice that we do not automatically go over
+      // a root node boundary.
+      const assignedSlot =
+        /** @type {?Node|undefined} */ (n[ASSIGNED_SLOT_PROP]) ||
+        n.assignedSlot;
+      if (assignedSlot) {
+        n = assignedSlot;
+      } else {
+        n = n.parentNode;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Direct slot assignment. Works the same way as shadow slots, but does not
+   * require a shadow root. Automatically starts the discovery phase for the
+   * affected nodes.
+   *
+   * See `Element.assignedSlot` API.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot to which the target node is assigned.
+   */
+  static assignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] == slot) {
+      return;
+    }
+    node[ASSIGNED_SLOT_PROP] = devAssert(slot);
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Unassigns the direct slot previously done by the `assignSlot` call.
+   * Automatically starts the discovery phase for the affected nodes.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot from which the target node is assigned.
+   */
+  static unassignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] != slot) {
+      return;
+    }
+    delete node[ASSIGNED_SLOT_PROP];
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Creates the context node and automatically starts the discovery process.
+   *
+   * @param {*} privateOnly
+   * @param {!Node} node
+   */
+  constructor(privateOnly, node) {
+    devAssert(privateOnly === PRIVATE_ONLY);
+
+    /** @const @private {!Node} */
+    this.node_ = node;
+
+    /** @private {?ContextNode} */
+    this.parent_ = null;
+
+    /** @private {boolean} */
+    this.parentOverriden_ = false;
+
+    /** @private {boolean} */
+    this.isRoot_ = node.nodeType == DOCUMENT_NODE;
+
+    /** @private {?ContextNode} */
+    this.root_ = this.isRoot_ ? this : null;
+
+    /** @private {?Array<!ContextNode>} */
+    this.children_ = null;
```

Comment:
Why not default to `[]`? It looks like that is what happens anyway in L315 and the test-only method.

Reply:
I expect that we will have a lot more terminal nodes than parent nodes. I.e. the tree will be branchy and wide. So, I wanted to save on extra allocation. For tests, of course, this optimization is not as important.